src/index.ts

```diff
@@ -85,16 +84,26 @@ const zodToTsNode = (
       return zodLiteralToTs(zod._def.value)
     case 'ZodObject': {
       const properties = Object.entries(zod._def.shape())
-
       const members: ts.TypeElement[] = properties.map(([key, value]) => {
         const zodAny = value as ZodTypeAny
-        const type = zodToTsNode(zodAny, ...otherArgs, { parentIsObject: true })
+        let type = zodToTsNode(zodAny, ...otherArgs) as ts.TypeNode
 
+        const optionalMember = zodAny._def.typeName === 'ZodOptional'
         const flagsRequireOptional = options.treatOptionalsAs === 'both' || options.treatOptionalsAs === 'optional'
+
+        if (
+          optionalMember && options.treatOptionalsAs === 'optional' && isUnionTypeNode(type)
+        ) {
+          // todo typings
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
+          type = f.updateUnionTypeNode(type, type.types.filter(p => p.kind !== ts.SyntaxKind.UndefinedKeyword))
```

Comment:
The idea of removing undefined here sounds good and more robust. This won't work though since the unions can be nested. I wonder if we can just apply a TS transform on the node returned from `zodToTs`.

Reply:
Can we safely flatten all the potentially unnecessary nesting? In this case we can - `a: (number | undefined) | string;`. I wonder if typescript has out of the box solution for this.

Reply:
Wasn't able to find anything in a simple search. Flattening would be great though.